clang/lib/AST/PreorderAST.cpp

```diff
@@ -145,34 +153,200 @@ void PreorderAST::Create(Expr *E, OperatorNode *Parent) {
+    }
+
+  } else if (auto *AE = dyn_cast<ArraySubscriptExpr>(E)) {
+    // e1[e2] has the same canonical form as *(e1 + e2).
+    auto *DerefExpr = BinaryOperator::Create(Ctx, AE->getBase(), AE->getIdx(),
+                                             BinaryOperatorKind::BO_Add, AE->getType(),
+                                             AE->getValueKind(), AE->getObjectKind(),
+                                             AE->getExprLoc(), FPOptionsOverride());
+    auto *N = new UnaryOperatorNode(UnaryOperatorKind::UO_Deref, Parent);
+    AddNode(N, Parent);
+    // Even though e1 + e2 is already a binary operator, the child of the
+    // UnaryOperatorNode should be e1 + e2 + 0. This enables expressions such
+    // as p[i + -(1 + 2)] to be constant folded. In order for an OperatorNode
+    // to be constant folded, it must have at least two LeafExprNode children
+    // whose expressions are integer constants. For example, i + -(1 + 2) + 0
+    // will be constant folded to i + -3, but i + -(1 + 2) will not be constant
+    // folded.
+    AddZero(DerefExpr, /*Parent*/ N);
+
+  } else if (auto *ICE = dyn_cast<ImplicitCastExpr>(E)) {
+    auto *N = new ImplicitCastNode(ICE->getCastKind(), Parent);
+    AddNode(N, Parent);
+    Create(ICE->getSubExpr(), /*Parent*/ N);
+
   } else {
     auto *N = new LeafExprNode(E, Parent);
     AddNode(N, Parent);
   }
 }
 
+void PreorderAST::AddZero(Expr *E, Node *Parent) {
+  auto *N = new OperatorNode(BO_Add, Parent);
+  AddNode(N, Parent);
+
+  llvm::APInt Zero(Ctx.getTargetInfo().getIntWidth(), 0);
+  auto *ZeroLiteral = new (Ctx) IntegerLiteral(Ctx, Zero, Ctx.IntTy,
+                                               SourceLocation());
+  auto *L = new LeafExprNode(ZeroLiteral, N);
+  AddNode(L, /*Parent*/ N);
+  Create(E, /*Parent*/ N);
+}
+
 void PreorderAST::Coalesce(Node *N, bool &Changed) {
   if (Error)
     return;
 
-  auto *O = dyn_cast_or_null<OperatorNode>(N);
-  if (!O)
+  if (!N)
     return;
 
-  // Coalesce the children first.
-  for (auto *Child : O->Children)
-    if (isa<OperatorNode>(Child))
-      Coalesce(Child, Changed);
+  switch (N->Kind) {
+    case Node::NodeKind::OperatorNode: {
+      auto *O = dyn_cast<OperatorNode>(N);
 
-  if (CanCoalesceNode(O)) {
-    CoalesceNode(O);
-    Changed = true;
+      // Coalesce the children first.
+      for (auto *Child : O->Children)
+        Coalesce(Child, Changed);
+
+      if (CanCoalesceNode(O)) {
+        CoalesceNode(O);
+        Changed = true;
+      }
+      break;
+    }
+    case Node::NodeKind::UnaryOperatorNode: {
+      auto *U = dyn_cast<UnaryOperatorNode>(N);
+      Coalesce(U->Child, Changed);
+      break;
+    }
+    case Node::NodeKind::MemberNode: {
+      auto *M = dyn_cast<MemberNode>(N);
+      Coalesce(M->Base, Changed);
+      break;
+    }
+    case Node::NodeKind::ImplicitCastNode: {
+      auto *I = dyn_cast<ImplicitCastNode>(N);
+      Coalesce(I->Child, Changed);
+      break;
+    }
+    default:
```

Comment:
Having `default` as the first case in a `switch` allows better code readability as the switch may become longer when new cases are added to it in future.

Also do we even need a default case here?

Reply:
The default case avoids a warning that would be emitted if the switch did not handle the `LeafExprNode` case.